File: sdk-python/cryptoagent/action.py
"""Canonical agent action message.

Mirror of go-key-service/internal/action/action.go. See docs/schema.md
for the contract; both libraries MUST produce byte-identical canonical
encodings for equal inputs.
"""

from __future__ import annotations

import json
from dataclasses import dataclass

SCHEMA_VERSION = 1
NONCE_HEX_LEN = 32
MAX_SKEW_MS = 30_000
NONCE_WINDOW_MS = 600_000

_HEX_LOWER = set("0123456789abcdef")


class ActionError(ValueError):
    """Base class for action validation errors."""


@dataclass(frozen=True)
class Action:
    schema_version: int
    agent_id: str
    action_type: str
    target: str
    timestamp_ms: int
    nonce: str

    def validate(self) -> None:
        if self.schema_version != SCHEMA_VERSION:
            raise ActionError(
                f"unsupported schema_version: got {self.schema_version} want {SCHEMA_VERSION}"
            )
        if not self.agent_id or not self.action_type or not self.target:
            raise ActionError("required field empty")
        if len(self.nonce) != NONCE_HEX_LEN or any(c not in _HEX_LOWER for c in self.nonce):
            raise ActionError("nonce must be 32 lowercase hex chars")
        if not isinstance(self.timestamp_ms, int) or isinstance(self.timestamp_ms, bool):
            raise ActionError("timestamp must be int milliseconds")

    def canonical(self) -> bytes:
        """Deterministic JSON bytes used for signing and hashing."""
        self.validate()
        payload = {
            "action_type": self.action_type,
            "agent_id": self.agent_id,
            "nonce": self.nonce,
            "schema_version": self.schema_version,
            "target": self.target,
            "timestamp": self.timestamp_ms,
        }
        return json.dumps(
            payload,
            sort_keys=True,
            separators=(",", ":"),
            ensure_ascii=False,
        ).encode("utf-8")

File: go-key-service/internal/action/action.go
// Package action defines the canonical agent action message and its
// deterministic encoding. See docs/schema.md for the full contract.
package action

import (
	"bytes"
	"encoding/hex"
	"encoding/json"
	"errors"
	"fmt"
)

const (
	SchemaVersion = 1
	NonceHexLen   = 32
	MaxSkewMillis = 30_000
	NonceWindowMs = 600_000
)

type Action struct {
	SchemaVersion int    `json:"schema_version"`
	AgentID       string `json:"agent_id"`
	ActionType    string `json:"action_type"`
	Target        string `json:"target"`
	TimestampMs   int64  `json:"timestamp"`
	Nonce         string `json:"nonce"`
}

var (
	ErrSchemaVersion = errors.New("action: unsupported schema_version")
	ErrEmptyField    = errors.New("action: required field empty")
	ErrNonceShape    = errors.New("action: nonce must be 32 lowercase hex chars")
)

// Validate checks structural invariants. Replay-window checks
// (timestamp skew, nonce reuse) are the verifier's job.
func (a *Action) Validate() error {
	if a.SchemaVersion != SchemaVersion {
		return fmt.Errorf("%w: got %d want %d", ErrSchemaVersion, a.SchemaVersion, SchemaVersion)
	}
	if a.AgentID == "" || a.ActionType == "" || a.Target == "" {
		return ErrEmptyField
	}
	if len(a.Nonce) != NonceHexLen {
		return ErrNonceShape
	}
	if _, err := hex.DecodeString(a.Nonce); err != nil {
		return ErrNonceShape
	}
	for _, r := range a.Nonce {
		if r >= 'A' && r <= 'F' {
			return ErrNonceShape
		}
	}
	return nil
}

// Canonical returns the deterministic JSON bytes used for signing
// and hashing. Keys are sorted; no insignificant whitespace.
func (a *Action) Canonical() ([]byte, error) {
	if err := a.Validate(); err != nil {
		return nil, err
	}
	// encoding/json sorts map keys. Marshal via map to guarantee order
	// independent of struct field declaration order.
	m := map[string]any{
		"schema_version": a.SchemaVersion,
		"agent_id":       a.AgentID,
		"action_type":    a.ActionType,
		"target":         a.Target,
		"timestamp":      a.TimestampMs,
		"nonce":          a.Nonce,
	}
	var buf bytes.Buffer
	enc := json.NewEncoder(&buf)
	enc.SetEscapeHTML(false)
	if err := enc.Encode(m); err != nil {
		return nil, err
	}
	// Encoder appends a trailing newline; strip it.
	out := buf.Bytes()
	if n := len(out); n > 0 && out[n-1] == '\n' {
		out = out[:n-1]
	}
	return out, nil
}
